plugins/embedding.cc

```diff
@@ -1334,6 +1336,269 @@ regNearestNeighborsFunction(builtinPackage(),
+/*****************************************************************************/
+
+DEFINE_STRUCTURE_DESCRIPTION(ImageWrapperFunctionConfig);
+
+ImageWrapperFunctionConfigDescription::
+ImageWrapperFunctionConfigDescription()
+{
+    addField("expression",
+             &ImageWrapperFunctionConfig::expression,
+             "");           
+}
+
+ImageWrapperInput::
+ImageWrapperInput()
+{
+}
+
+DEFINE_STRUCTURE_DESCRIPTION(ImageWrapperInput);
+
+ImageWrapperInputDescription::
+ImageWrapperInputDescription()
+{
+    addField("x", &ImageWrapperInput::x,
+             "");
+    addField("y", &ImageWrapperInput::y,
+             "");
+}
+
+DEFINE_STRUCTURE_DESCRIPTION(ImageWrapperOutput);
+
+ImageWrapperOutputDescription::
+ImageWrapperOutputDescription()
+{
+    addField("value", &ImageWrapperOutput::value,
+             "");
+}
+
+ImageWrapperFunction::
+ImageWrapperFunction(MldbServer * owner,
+                         PolyConfig config,
+                         const std::function<bool (const Json::Value &)> & onProgress)
+    : BaseT(owner)
+{
+    functionConfig = config.params.convert<ImageWrapperFunctionConfig>();
+    SqlExpressionMldbScope context(owner);
+    auto boundExpr = functionConfig.expression->bind(context);
+    SqlRowScope scope;
+    embedding = boundExpr(scope, GET_ALL);
+    shape = embedding.getEmbeddingShape();
+}
+
+ImageWrapperFunction::
+~ImageWrapperFunction()
+{
+}
+
+struct ImageWrapperFunctionApplier
+    : public FunctionApplierT<ImageWrapperInput, ImageWrapperOutput> {
+
+    ImageWrapperFunctionApplier(const Function * owner)
+        : FunctionApplierT<ImageWrapperInput, ImageWrapperOutput>(owner)
+    {
+        info = owner->getFunctionInfo();
+    }
+};
+
+ImageWrapperOutput
+ImageWrapperFunction::
+applyT(const ApplierT & applier_, ImageWrapperInput input) const
+{   
+    ImageWrapperOutput output;    
+
+    //We clamp, we do not currently provide interpolation
+    int x = input.x.coerceToInteger().toInt();
+    int y = input.y.coerceToInteger().toInt();
+    x = boost::algorithm::clamp(x, 0, shape[0]-1);
+    y = boost::algorithm::clamp(y, 0, shape[1]-1);
+
+    ColumnName columnName;
+    columnName = columnName + PathElement(y);
+    columnName = columnName + PathElement(x);
+
+    ExpressionValue storage;
+    auto pValue = embedding.tryGetNestedColumn(columnName, storage);
+
+    if (pValue)
+        return {*pValue};
+    else
+        return {ExpressionValue(0, Date::negativeInfinity())};
+}
+    
+std::unique_ptr<FunctionApplierT<ImageWrapperInput, ImageWrapperOutput> >
+ImageWrapperFunction::
+bindT(SqlBindingScope & outerContext, const std::shared_ptr<RowValueInfo> & input) const
+{
+    std::unique_ptr<ImageWrapperFunctionApplier> result
+        (new ImageWrapperFunctionApplier(this));
+ 
+    return std::move(result);
+}
+
+static RegisterFunctionType<ImageWrapperFunction, ImageWrapperFunctionConfig>
+regImageWrapperFunction(builtinPackage(),
+                            "image.wrapper",
```

Comment:
image.readPixels?

Also, the "right" way for this to work in SQL would be to have a dataset function `embedding_dataset()` just like `row_table()` that allowed it to be done as a subselect.  For later.